[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46057 [PyTorch] [BUCK] Replace pt_deps.bzl with a YAML operator dependency file which is generated by the code analyser
* **#45722 [RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction**

This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

dhruvbird committed:
[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction

This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

ljk53 commented on tools/codegen/selective_build/selector.py:
```diff
@@ -0,0 +1,148 @@
+        if name in self.operators:
+            op: SelectiveBuildOperator = self.operators[name]
+            return op.is_root_operator
+        name = strip_operator_overload_name(name)
+        if name not in self.operators:
+            return False
+        base_op: SelectiveBuildOperator = self.operators[name]
+        return base_op.include_all_overloads and base_op.is_root_operator
```
root ops should always include overload part, right? We shouldn't apply `include_all_overloads` wildcard here, should we?

I know it breaks the symmetry - we can add a comment (include_all_overloads is being deprecated anyway).

dhruvbird replied:
This one was a bit trick to reason about. Here's how I am thinking about it - which may be wrong.

Consider you have the following ops `aten::add`, `aten::add.int`, `aten::add.float`.
Now, the presence of `aten::add` could be 2 things.
1. The specific operator named `aten::add`, or
2. All overloads with `aten::add` as the base operator name

We have no way to distinguish between them except for the `include_all_overloads` flag. If `include_all_overloads=True`, then it means that `aten::add` stands for the family of operators whose base name is `aten::add`, whereas if `include_all_overloads=False`, then it means that this entry stands for exactly one operator that has the exact name `aten::add`.

In selective build, if `aten::add` is present with `include_all_overloads=False` and the question is asked about including `aten::add.double`, and that entry doesn't exist, we will fall back to checking if the base operator name `aten::add` exists. If it does, we need to additionally check if that operator name is [1] or [2] (above).

ljk53 commented on tools/codegen/selective_build/selector.py:
```diff
@@ -0,0 +1,148 @@
+        if self.include_all_operators:
+            return True
+
+        if name in self.operators:
+            op: SelectiveBuildOperator = self.operators[name]
+            return op.is_used_for_training
```
Is it possible to have the following setup?
entry 1: {name: "aten::add.Tensor", is_root_operator: true, is_used_for_training: false, include_all_overloads: false}
entry 2: {name: "aten::add", is_root_operator: false, is_used_for_training: true, include_all_overloads: true}

Entry 1 is from root operator list of a non-training use case. Entry 2 is from the transitive closure of a training use case.

With the current logic, Entry 1 takes precedence, thus it returns false; but it should be true, right?

dhruvbird replied:
@ljk53 Good point! I had overlooked this one - will add it to the unit test to make sure that this doesn't regress.
My assumption was that with Training ops, the concept of the transitive closure is non-functional, and we need to always enumerate the complete set of training ops, but that doesn't seem to be true.

I'm guessing this special case applies only to training operator check, and not for the check for `is_root_operator`.

dhruvbird replied:
Resolved in latest rev (with tests).

ljk53 commented on tools/setup_helpers/generate_code.py:
```diff
@@ -79,6 +77,85 @@ def generate_code(ninja_global=None,
+        #
+        # If we are in internal FB build, then we will expect one of the 2
+        # file paths below.
+        if "root_op_list.yaml" in selected_op_list_path or "combined_op_list.yaml" in selected_op_list_path:
+            is_root_operator = "root_op_list.yaml" in selected_op_list_path
+            is_used_for_training = "combined_op_list.yaml" in selected_op_list_path
```
For internal BUCK build, do we call this codegen multiple times to generate unboxing_wrapper, variable_kernels, respectively?
If for unboxing_wrapper pass, we only use the unboxing_wrapper output and ignore variable_kernels output, and for variable_kernels pass, we only use the variable_kernels output and ignore unboxing_wrapper output, then does it matter whether is_root_operator / is_used_for_training is set precisely?

Can we simply set both to true as the OSS setup?

dhruvbird replied:
+1 @ljk53  -  I think that this makes total sense - can merge the 2 - thanks for the simplification!

dhruvbird replied:
Merged.

dhruvbird commented on tools/codegen/selective_build/operator.py:
```diff
@@ -0,0 +1,154 @@
+    # TorchScript model, etc...). An operator is considered to be a
+    # root operator if it is called directly from any one of the models
+    # that this app was built for, so it may not be a root operator
+    # in all of the models that are used in this app.
```
TODO: Replace app with "instance of the pytorch library".

dhruvbird replied:
Replaced all instances.

dhruvbird commented on tools/codegen/selective_build/operator.py:
```diff
@@ -0,0 +1,154 @@
+    # is used for training in one or more models used by this app.
+    is_used_for_training: bool
+
+    # If True, it indicates that this operator instance (object) refers to an
+    # operator without the overload name and should apply to all overloads
+    # which have this operator name as the base name. This flag is applicable
+    # only for objects that have operator names without a DOT (period) character
+    # in them.
```
@ljk53 I'll the commentary here (about it going away in the future) since this is where the doc block for this field is.

dhruvbird commented on tools/codegen/selective_build/operator.py:
```diff
@@ -0,0 +1,154 @@
+        # an operator used for training in any of the models used
+        # in this app.
+        lhs.is_used_for_training or rhs.is_used_for_training,
+        lhs.include_all_overloads or rhs.include_all_overloads,
```
TODO: Add additional clarification about the lifetime of this flag in the doc-block.

dhruvbird replied:
Added.

dhruvbird commented on tools/codegen/gen.py:
```diff
@@ -194,8 +195,8 @@ def func(f: NativeFunction) -> Optional[str]:
-        if op_registration_whitelist is not None and \
-                f"aten::{f.func.name.name}" not in op_registration_whitelist and target is Target.REGISTRATION:
+        op_name = f"aten::{f.func.name}"
+        if target is Target.REGISTRATION and not selector.is_operator_selected(op_name):
```
@ljk53 Sounds good! I'm assuming you mean have `is_operator_selected` take in 2 aguments `(namespace_name, operator)` right?

dhruvbird replied:
I think using a helper to do this transformation will be useful.

dhruvbird replied:
Will send out a follow up diff for this change since it's largely internal and doesn't affect any public facing API.

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

ezyang commented on tools/codegen/gen.py:
```diff
@@ -952,6 +953,32 @@ def write_outputs(self, filename: str) -> None:
+    if provided_op_registration_allowlist is not None:
+        op_registration_allowlist = set(provided_op_registration_allowlist)
+
+    selector: SelectiveBuilder = SelectiveBuilder.get_nop_selector()
```
nit: just assign selector = get_no_selector in the else branch

dhruvbird replied:
Addressed in latest rev.

ezyang commented on tools/codegen/selective_build/selector.py:
```diff
@@ -0,0 +1,154 @@
+    include_all_operators: bool
+
+    # Debug Information at the selective/custom build level.
+    debug_info: Optional[List[str]]
```
Is it possible to get more descriptive types here (and in other parts) than just `str`? For example, operator names in the main model are represented as a dataclass, rather than as just strings.

Also, avoid putting mutable types like `List` inside frozen dataclasses, they let you backdoor mutation. Make them tuples instead.

dhruvbird replied:
> Is it possible to get more descriptive types here (and in other parts) than just `str`? For example, operator names in the main model are represented as a dataclass, rather than as just strings.

I did have something more fine grained in an earlier version of this change, but @ljk53 and I felt like having this be more opaque would be better since this is going to be part of OSS. Thoughts?

> 
> Also, avoid putting mutable types like `List` inside frozen dataclasses, they let you backdoor mutation. Make them tuples instead.

Oh! Thanks - I didn't know that List was mutable - thanks for pointing it out! I'll change it to `Tuple[str...]` or `Sequence[str]`?

ezyang replied:
> I did have something more fine grained in an earlier version of this change, but @ljk53 and I felt like having this be more opaque would be better since this is going to be part of OSS. Thoughts?

There is nothing wrong with being opaque, but then there should be other indicators (e.g., leading underscores) that indicate that the representation is opaque and shouldn't be leaking beyond the file.

> Oh! Thanks - I didn't know that List was mutable - thanks for pointing it out! I'll change it to Tuple[str...] or Sequence[str]?

Tuple is better, as Sequence will let you store lists as the internal rep which means that they won't be hashable.

dhruvbird replied:
> There is nothing wrong with being opaque, but then there should be other indicators (e.g., leading underscores) that indicate that the representation is opaque and shouldn't be leaking beyond the file.

Got it! I'll use underscores for things of this nature (fixed).

> Tuple is better, as Sequence will let you store lists as the internal rep which means that they won't be hashable.

Thanks! Changed to Tuple.

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird commented:
How to flag a PR as being ready for review? I don't see a "needs review" or "request changes" option here.

dhruvbird commented:
nvm, got it! There's a "re-request" review icon that looks like an arrow going around in a circle.

ljk53 commented on tools/autograd/utils.py:
```diff
@@ -74,6 +74,12 @@ def is_tensor_method(declaration):
 
+def op_name_with_overload(decl):
+    name = decl['operator_name_with_overload']
+    if is_out_variant(decl):
+        name = decl['operator_name_with_overload'][:-len('_out')]
```
This looks weird - 'operator_name_with_overload' is the main name + overload name string, while is_out_variant() is to check whether the main name ends with '_out' - so looks like it removes the last 4 chars from overload part if main name ends with '_out'...

Here is an example:
name: eye_out
overload_name: m_out
schema_string: aten::eye.m_out(int n, int m, *, Tensor(a!) out) -> Tensor(a!)

operator_name_with_overload would be 'aten::eye.m_out'
and this method would return 'aten::eye.m'

dhruvbird replied:
Good catch! I'm curious to know what the rationale behind removing the out-variant (stripping out the trailing `_out`) is here?

dhruvbird replied:
@ljk53 I'm seeing that in Declarations.yaml (by convention?) out variants have an overload name of `.out`. For example:

```
- name: abs_out
  operator_name: abs
  overload_name: out
  use_c10_dispatcher: with_codegenerated_unboxing_wrapper
  manual_kernel_registration: false
  category_override: ''
  matches_jit_signature: true
  schema_string: aten::abs.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
```

and:

```
- name: polar_out
  operator_name: polar
  overload_name: out
  use_c10_dispatcher: with_codegenerated_unboxing_wrapper
  manual_kernel_registration: false
  category_override: ''
  matches_jit_signature: true
  schema_string: aten::polar.out(Tensor abs, Tensor angle, *, Tensor(a!) out) -> Tensor(a!)
```

Q. Was the original intention to keep all the overloads with an `.out` suffix if the base operator is included in the selective build?

i.e. if `abc`, `bcd`, and `xyz` are selected, then also select `abc.out`, `bcd.out`, and `xyz.out`?

dhruvbird replied:
@ljk53 Okay - I think I can remove this logic altogether and just return the full operator name with overload, and it should be fine. lmk what you think.

ljk53 replied:
I think the main operator name always has "_out" suffix if it's an out-variant - but the overload part might not always be ".out", (eye_out is an counterexample, its overload is m_out). I think it's fine to apply [:-len('_out')] on the main name, but applying it on the concatenated name `operator_name_with_overload` seems confusing...

I'm not 100% sure about the original intention of this conversion - maybe if an op is selected, then its base and out-variant can both be called and should include both? We should dig into it later :)

dhruvbird replied:
> maybe if an op is selected, then its base and out-variant can both be called and should include both? We should dig into it later :)

I also think this is the main reason for doing so. **If** that is the case, then removing the check entirely should be fine because:
1. In case `include_all_overloads` is true, the build will pick up all overloads (even the ones with the `.out` variant)
2. When we switch to tracing for training, we will know all the operator names (with overload), and then we don't need to over-select ops just for training purposes. Till such time that we don't do tracing for training, [1] should cover the case - thoughts?

dhruvbird replied:
@ljk53 So for `name: eye_out`, there are 2 overloads, one with `out` and one with `m_out`, so previously when removing the `_out` suffix from base operator name, things seem fine, but doing this with the overload is clearly not okay like you mentioned. I believe `include_all_overloads=True` should include both overloads (which is what the earlier code was trying to do?)

ljk53 replied:
ok, I took a closer took - seems it's because the Declarations.yaml and native_functions.yaml have slightly different naming convention. In native functions.yaml, the out-place version and base version have the same 'name', e.g.:
```
- func: abs(Tensor self) -> Tensor
- func: abs.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
```

In Declarations.yaml, they have different 'name':
```
- name: abs
  operator_name: abs
  overload_name: ''
- name: abs_out
  operator_name: abs
  overload_name: out
```

The unboxing_wrapper codegen operates on Declarations.yaml, but the allowlist ops are in native functions 'name' format. E.g. root_ops = ['aten::abs.out"], after stripping out the namespace and the overload part, we get 'abs', which still won't match the `decl['name']` - I think that's why the old logic has to remove the '_out' suffix in the 'name' field.

`operator_name_with_overload` seems to be generated from native_functions.yaml schema string, i.e. the 'name' part doesn't have the '_out' suffix, so this step doesn't seem to be necessary.

dhruvbird replied:
@ljk53 I agree - and am wondering if it would have been simpler for the unboxing/autograd codegen to look at `operator_name` instead of `name` (since `operator_name` is what maps to the name in `native_functions.yaml`, and not `name`, which seems to be something else).

ljk53 reviewed (approved):
Looks great! Just one comment around 'op_name_with_overload' should be addressed before landing. Others are mostly nits.

dhruvbird commented on tools/codegen/selective_build/operator.py:
```diff
@@ -0,0 +1,159 @@
+    @staticmethod
+    def from_legacy_operator_name_without_overload(name: str) -> 'SelectiveBuildOperator':
+        return SelectiveBuildOperator(
+            name,
+            True,
+            True,
+            True,
+            None,
```
Thanks for pointing this out - in fact I think that all call-sites should pass in parameter names to prevent against bugs in case 2 boolean members are switched around. i.e. if that happens, mypy won't complain, but it will be a logical error.

dhruvbird commented on tools/codegen/selective_build/selector.py:
```diff
@@ -0,0 +1,154 @@
+        if self.include_all_operators:
+            return True
+
+        not_training_op = SelectiveBuildOperator('', False, False, False, None)
```
Fixed.

dhruvbird commented on tools/setup_helpers/generate_code.py:
```diff
@@ -79,6 +77,56 @@ def generate_code(ninja_global=None,
+
+
+def get_selector(selected_op_list_path, operators_yaml_path):
+    # cwrap depends on pyyaml, so we can't import it earlier
+    root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+    sys.path.insert(0, root)
```
@ljk53 I ran into a problem in a windows CI build (iirc) because the paths aren't set up correctly. In fact the comment above about `cwrap` doesn't make sense to me since it's not used in this file at all. I agree that this should be put into main. Let me send out another change set on top of this one to address it and clean it up.

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Switch PyTorch Selective Build (Custom Build) to use the SelectiveBuilder abstraction"


This diff does a bunch of things:

1. Introduces some abstractions as detailed in https://fb.quip.com/2oEzAR5MKqbD to help with selective build related codegen in multiple files.
2. Adds helper methods to combine operators, debug info, operator lists, etc...
3. Currently, the selective build machinery querying `op_registration_whitelist` directly at various places in the code. `op_registration_whitelist` is a list of allowed operator names (without overload name). We want to move to a world where the overload names are also included so that we can be more selective about which operators we include. To that effect, it makes sense to hide the checking logic in a separate abstraction and have the build use that abstraction instead of putting all this selective build specific logic in the code-generator itself. This change is attempting to do just that.
4. Updates generate_code, unboxing-wrapper codegen, and autograd codegen to accept the operator selector paradigm as opposed to a selected operator list.
5. Update `tools/code_analyzer/gen_op_registration_allowlist.py` to expose providing an actual structured operator dependency graph in addition to a serialized string.

There are a bunch of structural changes as well:

1. `root_ops_list.yaml` and `combined_ops_list.yaml` are now actual YAML files (not a space separated list of operator names)
2. `generate_code.py` accepts only paths to operator list YAML files (both old style as well as new style) and not list of operator names on the command line as arguments

Differential Revision: [D23837010](https://our.internmc.facebook.com/intern/diff/D23837010/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23837010/)!

[ghstack-poisoned]

gchanan commented:
Please post your RFCs publicly.  We even have a repo for them!  https://github.com/pytorch/rfcs.
